### PR TITLE
ARROW-7022, ARROW-7023: [Python] fix handling of pandas Index and Period/Interval extension arrays in pa.array

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Fixes https://issues.apache.org/jira/browse/ARROW-7022, and while doing this noticed another bug this is fixing (for which I opened https://issues.apache.org/jira/browse/ARROW-7023)

### CODE REVIEW

File: python/pyarrow/pandas-shim.pxi [hunk 1]
```diff
@@ -183,6 +186,28 @@ cdef class _PandasAPIShim(object):
         else:
             return False
 
+    cpdef is_index(self, obj):
+        if self._have_pandas_internal():
```
Comment (pitrou) [inline python/pyarrow/pandas-shim.pxi]: I'm curious, why is this necessary? Perhaps add a comment?

File: python/pyarrow/tests/test_pandas.py [hunk 2]
```diff
@@ -3222,6 +3222,47 @@ def test_array_protocol():
...
+
+
+def PandasArray__arrow_array__(self, type=None):
+    # harcode dummy return regardless of self - we only want to check that
+    # this method is correctly called
+    storage = pa.array([1, 2, 3], type=pa.int64())
+    return pa.ExtensionArray.from_storage(DummyExtensionType(), storage)
+
+
+def test_array_protocol_pandas_extension_types(monkeypatch):
```
Comment (pitrou) [inline python/pyarrow/tests/test_pandas.py]: I wonder when `pytest` will marry `web2py`...

**Review by pitrou — COMMENTED** [review]
LGTM on the principle.

**Comment by pitrou** [comment]
I've restarted AppVeyor as it seems the build had stopped abruptly.

File: python/pyarrow/pandas-shim.pxi (hunk 1 above)
Comment (jorisvandenbossche) [reply python/pyarrow/pandas-shim.pxi]: All the other `is_..` methods above use the same pattern. 
The reason it is needed is because this method can in principle be called in a situation that pandas is not installed. So therefore we first check that with `self._have_pandas_internal` (the "internal" version is a cdef), and otherwise always return False.

Comment (pitrou) [reply python/pyarrow/pandas-shim.pxi]: Thanks for the explanation.

**Review by pitrou — APPROVED** [review]
+1

**Comment by jorisvandenbossche** [comment]
Thanks!
